src/RobustAdaptiveMetropolis.jl

```diff
@@ -0,0 +1,206 @@
+
+function step_inner(
+    rng::Random.AbstractRNG,
+    model::AbstractMCMC.LogDensityModel,
+    sampler::RAM,
+    state::RAMState
+)
+    # This is the initial state.
+    f = model.logdensity
+    d = LogDensityProblems.dimension(f)
+
+    # Sample the proposal.
+    x = state.x
+    U = randn(rng, d)
+    x_new = x + state.S * U
+
+    # Compute the acceptance probability.
+    lp = state.logprob
+    lp_new = LogDensityProblems.logdensity(f, x_new)
+    logα = min(lp_new - lp, zero(lp))  # `min` because we'll use this for updating
+    # TODO: use `randexp` instead.
+    isaccept = log(rand(rng)) < logα
+
+    return x_new, lp_new, U, logα, isaccept
+end
+
+function adapt(sampler::RAM, state::RAMState, logα::Real, U::AbstractVector)
+    # Update `
+    Δα = exp(logα) - sampler.α
+    S = state.S
+    # TODO: Make this configurable by defining a more general path.
+    η = state.iteration^(-sampler.γ)
+    ΔS = η * abs(Δα) * S * U / norm(U)
+    # TODO: Maybe do in-place and then have the user extract it with a callback if they really want it.
+    S_new = if sign(Δα) == 1
+        # One rank update.
+        LinearAlgebra.lowrankupdate(Cholesky(S), ΔS).L
+    else
+        # One rank downdate.
+        LinearAlgebra.lowrankdowndate(Cholesky(S), ΔS).L
+    end
+    return S_new, η
+end
+
+function AbstractMCMC.step(
+    rng::Random.AbstractRNG,
+    model::AbstractMCMC.LogDensityModel,
+    sampler::RAM;
+    initial_params=nothing,
+    kwargs...
+)
+    # This is the initial state.
+    f = model.logdensity
+    d = LogDensityProblems.dimension(f)
+
+    # Initial parameter state.
+    x = initial_params === nothing ? rand(rng, d) : initial_params
+    # Initialize the Cholesky factor of the covariance matrix.
+    S = LowerTriangular(sampler.S === nothing ? diagm(0 => ones(eltype(sampler.γ), d)) : sampler.S)
+
+    # Constuct the initial state.
+    lp = LogDensityProblems.logdensity(f, x)
+    state = RAMState(x, lp, S, 0.0, 0, 1, true)
+
+    return AdvancedMH.Transition(x, lp, true), state
+end
+
+function AbstractMCMC.step(
+    rng::Random.AbstractRNG,
+    model::AbstractMCMC.LogDensityModel,
+    sampler::RAM,
+    state::RAMState;
+    kwargs...
+)
+    # Take the inner step.
+    x_new, lp_new, U, logα, isaccept = step_inner(rng, model, sampler, state)
+    # Accept / reject the proposal.
+    state_new = RAMState(isaccept ? x_new : state.x, isaccept ? lp_new : state.logprob, state.S, logα, state.η, state.iteration + 1, isaccept)
+    return AdvancedMH.Transition(state_new.x, state_new.logprob, state_new.isaccept), state_new
+end
+
+function valid_eigenvalues(S, lower_bound, upper_bound)
+    # Short-circuit if the bounds are the default.
+    (lower_bound == 0 && upper_bound == Inf) && return true
+    # Note that this is just the diagonal when `S` is triangular.
+    eigenvals = LinearAlgebra.eigenvals(S)
```

Comment:
Is this often non-triangular? If so, probably in a lot of cases we could do a quick check using e.g. Gershgorin disk theorem, which might enable us to short-circuit in O(d^2) in a lot of situations, rather than do a full eigenvalue solve in O(d^3). EDIT: Just read more of the code, and I guess `S` is always lower triangular and this is moot?

Reply:
> Just read more of the code, and I guess S is always lower triangular and this is moot?

Exactly :+1: